### PR TITLE
Added libpcsclite as a package

### PR DESCRIPTION
**Pull request by thewchan**
Checklist
- [ ] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [ ] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [ ] Source is from official source.
- [ ] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [ ] If static libraries are linked in, the license of the static library is packaged.
- [ ] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [ ] Build number is 0.
- [ ] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [ ] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [ ] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

**Comment by thewchan** [comment]
@BastianZim here. I'm particular unsure what I should do for testing

**Comment by BastianZim** [comment]
Often, there is just a make check that's being run or upstream includes tests. 

I would see if you can add anything from here, as that is how upstream builds it: https://salsa.debian.org/rousseau/PCSC/-/blob/master/.github/workflows/build.yml

File: recipes/libpcsclite/meta.yaml
Comment (BastianZim) [inline recipes/libpcsclite/meta.yaml:42]:
```suggestion
  license_file: 
    - GPL-3.0.txt
    - COPYING
```
There are some more here that probably need to be linked: https://github.com/LudovicRousseau/PCSC/blob/master/COPYING

Comment (thewchan) [reply recipes/libpcsclite/meta.yaml:42]: The text of all the linked license is in COPYING though, right?

Comment (BastianZim) [reply recipes/libpcsclite/meta.yaml:42]: Ahh true yes, read to quickly. Just including Copying and updating the license field is enough.

**Comment by thewchan** [comment]
@BastianZim here. I'm particular unsure what I should do for testing

> Often, there is just a make check that's being run or upstream includes tests.
> 
> I would see if you can add anything from here, as that is how upstream builds it: https://salsa.debian.org/rousseau/PCSC/-/blob/master/.github/workflows/build.yml

Got it... looks like I'd need to also figure out what source files to include for test?

**Comment by BastianZim** [comment]
Yes but sometimes the test is included in the build script because of what's needed to execute the tests. I would first focus on getting this built and then look at the tests later.

**Comment by thewchan** [comment]
@BastianZim hmm, looks like the build worked?

**Comment by thewchan** [comment]
@BastianZim @conda-forge/staged-recipes I'm able to get OSX build working, but the test doesn't work on that environment for some reason; something about "not being a git environment". Relatedly, I had to run the test during the build stage because I think otherwise I'll have to take the entire source distribution into the test environment, build again, and run the test there. Any pointers and review welcomed. Thanks!

**Comment by BastianZim** [comment]
That is outside my area of expertise but maybe someone else can help.

**Comment by thewchan** [comment]
@conda-forge/staged-recipes @conda-forge/help-c-cpp review requested

**Comment by carterbox** [comment]
> something about "not being a git environment"

Can't help without the actual error message.

File: recipes/libpcsclite/meta.yaml
Comment (carterbox) [inline recipes/libpcsclite/meta.yaml:44]: This section should have existence tests for each of the artifact types: config files, libraries, headers, binaries.

Binaries should be tested by calling them with `--help` or `--version`.

File: recipes/libpcsclite/meta.yaml
Comment (carterbox) [inline recipes/libpcsclite/meta.yaml:38]:
```suggestion
requirements:
  build:
    - {{ compiler('c') }}
    - {{ cdt('systemd-devel') }}  # [linux]
    - autoconf-archive
    - automake
    - flex
    - libtool
    - make
    - pkg-config 
    - sysroot_linux-64 2.17  # [linux64]
  host:
    - libusb
    - libudev  # [linux]
    - libudev-devel-cos6-x86_64  # [not linux]
```

Probably this package depends on libusb and libudev as libraries (not build tools) so they should go in host.

File: recipes/libpcsclite/meta.yaml
Comment (carterbox) [inline recipes/libpcsclite/meta.yaml:20]: I don't think you are actually installing anything to $PREFIX. You're just building and testing. Probably you need to call `make install`.

**Review by carterbox — CHANGES_REQUESTED** [review]

File: recipes/libpcsclite/meta.yaml
Comment (carterbox) [inline recipes/libpcsclite/meta.yaml:12]: There should also be a run_export section which pins this library according to it's ABI compatability.

**Review by carterbox — CHANGES_REQUESTED** [review]

File: recipes/libpcsclite/meta.yaml
Comment (thewchan) [reply recipes/libpcsclite/meta.yaml:44]: I had to run the test suite in the build section because I can't figure out how do to so otherwise, any pointers how I might proceed?

File: recipes/libpcsclite/meta.yaml
Comment (thewchan) [reply recipes/libpcsclite/meta.yaml:38]: let me try this.

**Comment by thewchan** [comment]
@BastianZim @carterbox thanks for your help; specifically @carterbox I will try to resolve your comments one at a time.

File: recipes/libpcsclite/meta.yaml
Comment (thewchan) [reply recipes/libpcsclite/meta.yaml:20]: @carterbox I'll try calling make install instead of make? And see what happens

Comment (thewchan) [reply recipes/libpcsclite/meta.yaml]: @carterbox I might need to use certain options to make install into $PREFIX, right now it's trying to install into /usr/bin?

File: recipes/libpcsclite/meta.yaml
Comment (leofang) [inline recipes/libpcsclite/meta.yaml:9]: Why don't we use the official releases as listed in https://pcsclite.apdu.fr/#download?

File: recipes/libpcsclite/meta.yaml
Comment (leofang) [inline recipes/libpcsclite/meta.yaml:17]: According to the official website, we should not build on macOS:
> Apple reimplemented PC/SC above their own CryptoTokenKit API.
> You should not need to build pcsc-lite on macOS.

https://pcsclite.apdu.fr/#supported

In fact that's what I see too with `openconnect` (I am the maintainer of openconnect-feedstock): On macOS, the Yubikey support (which would rely on pcsc on Linux) is built and enabled by default, without any effort.

File: recipes/libpcsclite/meta.yaml
Comment (leofang) [inline recipes/libpcsclite/meta.yaml:36]: ditto, this can be removed.

File: recipes/libpcsclite/meta.yaml
Comment (leofang) [reply recipes/libpcsclite/meta.yaml:44]: just add this
```suggestion
    - pcscd --version
```

File: recipes/libpcsclite/meta.yaml
Comment (leofang) [inline recipes/libpcsclite/meta.yaml:57]: Could you add me plz? 🙂
```suggestion
    - thewchan
    - leofang
```

**Review by leofang — CHANGES_REQUESTED** [review]
Appreciate for the effort @thewchan! This PR shows up promptly just when I started looking into it. I left some comments. The main one is macOS builds should be skipped.

File: recipes/libpcsclite/meta.yaml
Comment (carterbox) [reply recipes/libpcsclite/meta.yaml:44]: You can run the test suite during the build phase (or not at all). The test phase of this recipe is for packaging tests (check that the artifacts are packaged correctly) that's why we want existence tests and checks that the CLIs of any binaries work.

File: recipes/libpcsclite/meta.yaml
Comment (carterbox) [reply recipes/libpcsclite/meta.yaml]: Yes, you need to install the build artifacts to $PREFIX.

Comment (thewchan) [reply recipes/libpcsclite/meta.yaml]: Right but... how do I do that?

Comment (leofang) [reply recipes/libpcsclite/meta.yaml]: I think when you call `./configure` you need to pass `--prefix=$PREFIX`, and then `make install` will work.

File: recipes/libpcsclite/meta.yaml
Comment (thewchan) [reply recipes/libpcsclite/meta.yaml:48]: I can point it back, I think at some point I was using something like that but @BastianZim advised me to switch to GH

File: recipes/libpcsclite/meta.yaml
Comment (thewchan) [reply recipes/libpcsclite/meta.yaml:9]: I can do this as well, and I think again as per @BastianZim I used the GH instead but yeah I can point it here instead

Comment (leofang) [reply recipes/libpcsclite/meta.yaml:9]: Ah let's wait for @BastianZim's reply before any change, then. It's unclear to me why we need this. The CF bot can detect new version updates even if the source is not from GH.

File: recipes/libpcsclite/meta.yaml
Comment (thewchan) [reply recipes/libpcsclite/meta.yaml:44]: I'll do the CLI test that @leofang suggested thanks!

Comment (thewchan) [reply recipes/libpcsclite/meta.yaml:44]: @carterbox if you have other suggestions please let me know as I'm kinda clueless...

File: recipes/libpcsclite/meta.yaml
Comment (thewchan) [reply recipes/libpcsclite/meta.yaml]: It worked! Thanks

File: recipes/libpcsclite/meta.yaml
Comment (thewchan) [reply recipes/libpcsclite/meta.yaml:12]: Do you have more guidance about this part? I'm unfamiliar

File: recipes/libpcsclite/meta.yaml
Comment (thewchan) [reply recipes/libpcsclite/meta.yaml:9]: I'll go ahead and change it for now, easy enough to revert if needed

File: recipes/libpcsclite/meta.yaml
Comment (thewchan) [reply recipes/libpcsclite/meta.yaml]: Got it, do you know if Win build is impossible/inappropriate? I sort of assumed based on what was required

**Comment by thewchan** [comment]
@carterbox @leofang this is new, now it's complaining about `libtool` version I think? The version is stuck upstream right now waiting to be merged: https://github.com/conda-forge/libtool-feedstock/pull/29

File: recipes/libpcsclite/meta.yaml
Comment (leofang) [reply recipes/libpcsclite/meta.yaml:12]: According to the ABI lab (which is outdated) https://abi-laboratory.pro/index.php?view=timeline&l=pcsc-lite it's mostly backward compatible, so I guess this is enough:
```suggestion
build:
  run_exports:
    - {{ pin_subpackage('libpcsclite') }}
```
This ensures any packages built against `libpcsclite` v1.x will use v1.x.

File: recipes/libpcsclite/meta.yaml
Comment (leofang) [reply recipes/libpcsclite/meta.yaml:52]: ^^^^^^ @thewchan ?

File: recipes/libpcsclite/meta.yaml
Comment (leofang) [reply recipes/libpcsclite/meta.yaml:43]: ditto, not sure why GH is preferred in this case

File: recipes/libpcsclite/meta.yaml
Comment (thewchan) [reply recipes/libpcsclite/meta.yaml:52]: Yes sorry I will add this as well

**Comment by leofang** [comment]
@thewchan the problem is not really in the package source, but in the `bootstrap` script. The script was not sufficient, it should have `--force` added to the last line when calling `autoreconf`:
https://salsa.debian.org/rousseau/PCSC/-/blob/master/bootstrap#L17
It'd be nice if you regenerate the configuration this way, skipping `./bootstrap`, since the green light with the GH artifact may be a false positive.

**Comment by thewchan** [comment]
> @thewchan the problem is not really in the package source, but in the `bootstrap` script. The script was not sufficient, it should have `--force` added to the last line when calling `autoreconf`: https://salsa.debian.org/rousseau/PCSC/-/blob/master/bootstrap#L17 It'd be nice if you regenerate the configuration this way, skipping `./bootstrap`, since the green light with the GH artifact may be a false positive.

Okay, let me try using those lines verbatim instead of `./bootstrap`.

**Comment by thewchan** [comment]
@leofang it doesn't seem like the CLI test is working, either I didn't build the entry points correctly or there's something else amiss: https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=632472&view=logs&j=6f142865-96c3-535c-b7ea-873d86b887bd&t=22b0682d-ab9e-55d7-9c79-49f3c3ba4823&l=2492

**Comment by thewchan** [comment]
@leofang I'm checking the logs to make sure it installed properly in `$PREFIX`, I think at least it tried to? https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=632472&view=logs&j=6f142865-96c3-535c-b7ea-873d86b887bd&t=22b0682d-ab9e-55d7-9c79-49f3c3ba4823&l=1198

**Comment by thewchan** [comment]
Maybe I need to so something like `mv $PREFIX/sbin/* $PREFIX/bin/`? I see that in the wild. I'll try it

**Comment by thewchan** [comment]
> Maybe I need to so something like `mv $PREFIX/sbin/* $PREFIX/bin/`? I see that in the wild. I'll try it

https://github.com/conda-forge/slurm-feedstock/blob/3ecf889c74e52259b31e24ec58f5f6f518835be4/recipe/build.sh#L3

**Comment by thewchan** [comment]
@carterbox @leofang Okay it runs now! It possibly needs more test, and I also wonder if a Win build is possible. But thanks so much for your help!

File: recipes/libpcsclite/meta.yaml
Comment (leofang) [reply recipes/libpcsclite/meta.yaml]: I don't think this works on Windows. Even if it does, we can create a windows build later.

File: recipes/libpcsclite/meta.yaml
Comment (leofang) [inline recipes/libpcsclite/meta.yaml:20]: I noticed the build log says libusb support is not built, and I think this is why:
```suggestion
    - ./configure --prefix=$PREFIX --enable-libusb
```

File: recipes/libpcsclite/meta.yaml
Comment (leofang) [inline recipes/libpcsclite/meta.yaml:23]: This seems fragile, let me see if there's a better way.

File: recipes/libpcsclite/meta.yaml
Comment (leofang) [inline recipes/libpcsclite/meta.yaml:38]: I noticed `configure` tried to call `/usr/bin/file` and failed:
```suggestion
    - file
    - sysroot_linux-64 2.17  # [linux64]
```

**Comment by thewchan** [comment]
@leofang it seems that we can't build with libudev and libusb at the same time, we have to pick one:
 https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=632569&view=logs&j=6f142865-96c3-535c-b7ea-873d86b887bd&t=22b0682d-ab9e-55d7-9c79-49f3c3ba4823&l=1158

**Comment by leofang** [comment]
Yeah... I saw that too... Did a quick search, which to pick doesn't seem important (it is the default of PCSCLite, and both support hot plug), so how about just sticking to what we know works (`libudev`)?

File: recipes/libpcsclite/meta.yaml
Comment (leofang) [inline recipes/libpcsclite/meta.yaml:20]: Remove `libusb` build and fix the sbin folder:
```suggestion
    - ./configure --prefix=$PREFIX --sbindir=$PREFIX/bin
```

File: recipes/libpcsclite/meta.yaml
Comment (leofang) [reply recipes/libpcsclite/meta.yaml:23]: See above comment, this is no longer needed.
```suggestion
```

File: recipes/libpcsclite/meta.yaml
Comment (leofang) [inline recipes/libpcsclite/meta.yaml:42]: ditto
```suggestion
```

**Review by leofang — COMMENTED** [review]
btw @thewchan A quick tip in case you don't know already: You can just apply suggestions in batch on the GH web interface,  so you can quickly do the work everywhere (cell phone, laptop, ...) without editing in your local git repo 🙂

**Comment by thewchan** [comment]
@carterbox @leofang We're back to passing! Any other tests we should run?

**Review by leofang — APPROVED** [review]
LGTM! Thanks @thewchan! 

@carterbox @conda-forge/help-c-cpp This is ready.

File: recipes/libpcsclite/meta.yaml
Comment (leofang) [inline recipes/libpcsclite/meta.yaml:45]: Maybe one more test?
```suggestion
    - pcscd --version
    - test -f $PREFIX/include/pcsclite.h
```

File: recipes/libpcsclite/meta.yaml
Comment (carterbox) [inline recipes/libpcsclite/meta.yaml:25]: Most of the time ABI compatability does not sync with the API version. Do you have any documentation that they only break the ABI at major API releases?

Ask upstream whether they only do ABI breaks at certain API releases.

```suggestion
    - {{ pin_subpackage('libpcsclite', max_pin='x.x.x') }}
```

File: recipes/libpcsclite/meta.yaml
Comment (carterbox) [inline recipes/libpcsclite/meta.yaml:1]:
```suggestion
{% set name = "pcsclite" %}
```
This package also contains binaries with CLI. libpcsclite would be the subpackage of this recipe with only the library/headers. You can split the package into multiple outputs now if you want, but it's not necessary.

**Review by carterbox — CHANGES_REQUESTED** [review]

File: recipes/libpcsclite/meta.yaml
Comment (carterbox) [inline recipes/libpcsclite/meta.yaml:46]:
```suggestion
    - pcsc-spy --help
```

File: recipes/libpcsclite/meta.yaml
Comment (thewchan) [reply recipes/libpcsclite/meta.yaml:25]: @leofang What do  you think?

Comment (leofang) [reply recipes/libpcsclite/meta.yaml:25]: @carterbox i think it'd be too strict, see my earlier comment: https://github.com/conda-forge/staged-recipes/pull/21405#discussion_r1054838829 that said I would not mind if you wanna be on the cautious side

Comment (leofang) [reply recipes/libpcsclite/meta.yaml:25]: btw

> Most of the time ABI compatability does not sync with the API version.

I am not sure this is true for traditional Linux packages...

File: recipes/libpcsclite/meta.yaml
Comment (leofang) [reply recipes/libpcsclite/meta.yaml:1]: Yes indeed.

File: recipes/libpcsclite/meta.yaml
Comment (carterbox) [reply recipes/libpcsclite/meta.yaml:25]: > @carterbox i think it'd be too strict, see my earlier comment: [#21405 (comment)](https://github.com/conda-forge/staged-recipes/pull/21405#discussion_r1054838829) that said I would not mind if you wanna be on the cautious side

All of the samples on ABI lab are v1.8, so there is not enough evidence to say whether they would break the ABI at a minor API change. I think the breaking ABI changes from 1.8.9 to 1.8.10 were an accident. I still recommend asking upstream, but pinning to `'x.x'` is probably? safe.

**Comment by leofang** [comment]
I feel pcsc-spy should be temporarily removed from the package. It should be made a stand-alone Python noarch package that depends on libpcsclite, but since we have not set it up, it'd take a while to do. I'd prefer to not have it for now.

File: recipes/libpcsclite/meta.yaml
Comment (leofang) [reply recipes/libpcsclite/meta.yaml:45]: @carterbox good catch for adding pcsc-spy, I don't even know it exists...

File: recipes/libpcsclite/meta.yaml
Comment (leofang) [reply recipes/libpcsclite/meta.yaml:25]: Sounds good to me!

**Comment by thewchan** [comment]
Getting this now `/home/conda/staged-recipes/build_artifacts/libpcsclite_1671750348569/test_tmp/run_test.sh: /home/conda/staged-recipes/build_artifacts/libpcsclite_1671750348569/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_plac/bin/pcsc-spy: /usr/bin/python3: bad interpreter: No such file or directory`

**Comment by leofang** [comment]
Yes that's what I meant. We got this error because Python was not installed in the test environment (it's not a dependency as we missed it). I didn't know `pcsc-spy` exists. It's a simple Python script that can be invoked from cmdline. But then it means Python becomes a required run dependency for `pcsclite`, which is not ideal.

File: recipes/libpcsclite/meta.yaml
Comment (leofang) [reply recipes/libpcsclite/meta.yaml:25]:
```suggestion
    - {{ pin_subpackage('libpcsclite', max_pin='x.x') }}
```

**Comment by thewchan** [comment]
@carterbox @leofang okay I think i've made the requested change, name of package and removal of pcsc-spy.

File: recipes/libpcsclite/meta.yaml
Comment (leofang) [inline recipes/libpcsclite/meta.yaml:22]:
```suggestion
    - make distcheck
    - rm $PREFIX/bin/pcsc-spy
    - rm $PREFIX/lib/libpcscspy.so*
```

File: recipes/libpcsclite/meta.yaml
Comment (leofang) [inline recipes/libpcsclite/meta.yaml:49]:
```suggestion
```

File: recipes/libpcsclite/meta.yaml
Comment (leofang) [inline recipes/libpcsclite/meta.yaml:50]: @carterbox why do we care about this? It has the host prefix hard-coded which is useless.

File: recipes/libpcsclite/meta.yaml
Comment (leofang) [inline recipes/libpcsclite/meta.yaml:33]: I struggled to understand why this is still not used, but couldn't figure it out, but it seems no harm.
```suggestion
```

**Review by leofang — CHANGES_REQUESTED** [review]
@thewchan I think removing the test for `pcsc-spy` is not enough. We need to remove it completely.

This is how `pcsc-spy` works: 
- the user do `LD_PRELOAD=libpcscspy.so` which wraps all symbols from `libpcsclite.so` to dump a log to disk
- then, calling `pcsc-spy` separately would launch workers to read the log file

Basically, it is a log monitor that does not provide real functionality. To keep things cleanly separated we should not include the relevant files until we offer a separate subpackage (that depends on Python).

File: recipes/libpcsclite/meta.yaml
Comment (carterbox) [reply recipes/libpcsclite/meta.yaml:50]: Generally checking for CMake/Package config files indicates that they are being shipped without needing to download the package and look. Downstream builds can break if they depend on config files which are here today and removed later.

What do you mean useless? [Paths like `/home/conda/staged-recipes/build_artifacts/pcsclite_1672093677527/_h_env_placehold_placehold_placehold_` are replaced by conda with `$CONDA_PREFIX` at install-time.](https://docs.conda.io/projects/conda-build/en/latest/resources/make-relocatable.html)

**Comment by thewchan** [comment]
@carterbox @leofang I'll let you all duke it out I don't mind one way or another re: the tests. Just let me know.

File: recipes/libpcsclite/meta.yaml
Comment (leofang) [reply recipes/libpcsclite/meta.yaml:50]: You're right, I was being silly about the placeholder prefix. 

Still, I rarely see recipes check for .pc files (plz don't mix/compare them with cmake files; they are of vastly different values) and I can raise many counter examples, so if you think it's the standard practice it'd be better to document it.

**Review by leofang — APPROVED** [review]

**Comment by leofang** [comment]
Friendly nudge @carterbox 🙂

File: recipes/libpcsclite/meta.yaml
Comment (carterbox) [reply recipes/libpcsclite/meta.yaml:50]: I'm not commenting on the relative merits of CMake vs Package Config

If upstream developers of this package want to use PC, then we should ship them too. Testing whether they are there indicates to downstream users that we are including all of upstream's artifacts.

**Review by carterbox — CHANGES_REQUESTED** [review]

**Comment by thewchan** [comment]
Wow can't believe this is done! Thank you so much @carterbox and @leofang for your help!

**Pull request merged**
